// ngx-segment-analytics/esm2015/ngx-segment-analytics.module.js
import { NgModule, Optional, PLATFORM_ID, SkipSelf } from '@angular/core';
import { CommonModule, isPlatformBrowser } from '@angular/common';
import { SegmentService } from './ngx-segment-analytics.service';
import { SEGMENT_CONFIG } from './ngx-segment-analytics.config';
import { WindowWrapper } from './window-wrapper';
/**
 * Window Provider for Angular AOT
 * @returns Browser Window instance
 */
export function getWindow(platformId) {
    return isPlatformBrowser(platformId) ? window : {};
}
/**
 * Segment Module
 */
export class SegmentModule {
    /**
     * Segment Module Constructor
     *
     * @param parentModule Must be null
     */
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('SegmentModule is already loaded. Import it in the AppModule only');
        }
    }
    /**
     * Segment Module Initialisation
     *
     * @param config Segment Configuration
     * @returns Segment Module
     */
    static forRoot(config) {
        return {
            ngModule: SegmentModule,
            providers: [
                { provide: SEGMENT_CONFIG, useValue: config },
                SegmentService,
            ],
        };
    }
}
SegmentModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                providers: [
                    { provide: WindowWrapper, useFactory: getWindow, deps: [PLATFORM_ID] },
                ]
            },] }
];
SegmentModule.ctorParameters = () => [
    { type: SegmentModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
